os/kernel/binary_manager/binary_manager_load.c

```diff
@@ -91,17 +132,32 @@ static int binary_manager_read_header(int bin_idx, int part_idx, binary_header_t
 		return ERROR;
 	}
 
+	/* Read the checksum */
+	ret = read(fd, (FAR uint8_t *)&crc_hash, sizeof(uint32_t));
 	/* Read the binary header */
 	ret = read(fd, (FAR uint8_t *)header_data, sizeof(binary_header_t));
+
+	/* Verify header data */
+	if (binary_manager_verify_header(header_data) == ERROR) {
+		close(fd);
+		return ERROR;
+	}
+
+	file_size = header_data->header_size + header_data->bin_size;
+
+	if (binary_manager_verify_checksum(fd, file_size, crc_hash) == ERROR) {
+		close(fd);
+		return ERROR;
+	}
+
 	close(fd);
 
 	if (ret != sizeof(binary_header_t)) {
```

Comment:
We need to check whether reading succeed first before verifying header data.
So this should be line 139